File: ios/AICashier/Sources/Data/Receipt.swift
import Foundation

/// The receipt in the two forms Thai law knows (`server/services/receipt.py`):
/// an abbreviated tax invoice for a VAT-registered shop, a plain receipt for
/// everyone else.  32 columns, what a thermal printer takes as-is.
enum Receipt {
    static let width = 32

    private static func line(_ left: String, _ right: String = "") -> String {
        if right.isEmpty { return String(left.prefix(width)) }
        let l = String(left.prefix(width - right.count - 1))
        return l.padding(toLength: width - right.count, withPad: " ", startingAt: 0) + right
    }

    private static func centre(_ text: String) -> String {
        let t = String(text.prefix(width))
        let pad = max(0, (width - t.count) / 2)
        return String(repeating: " ", count: pad) + t
    }

    static func render(_ sale: Sale, settings: ShopSettings) -> String {
        let vat = settings.vatRegistered
        let f = DateFormatter()
        f.dateFormat = "dd/MM/yyyy HH:mm"
        var out = [centre(settings.storeName)]
        if !settings.storeAddress.isEmpty { out.append(centre(settings.storeAddress)) }
        if vat {
            out += [centre("ใบกำกับภาษีอย่างย่อ"), centre("TAX INVOICE (ABB)"),
                    centre("TIN \(settings.tin.isEmpty ? "-" : settings.tin)")]
        } else {
            out.append(centre("ใบเสร็จรับเงิน / RECEIPT"))
        }
        out += [line("No. \(sale.id)"), line(f.string(from: sale.timestamp)), String(repeating: "-", count: width)]
        for it in sale.items {
            out.append(line(it.productName))
            out.append(line("  \(it.quantity) x " + String(format: "%.2f", it.price), String(format: "%.2f", it.total)))
        }
        out.append(String(repeating: "-", count: width))
        if vat {
            out.append(line("Subtotal", String(format: "%.2f", sale.subtotal)))
            out.append(line("VAT \(Int((settings.taxRate * 100).rounded()))%", String(format: "%.2f", sale.tax)))
        }
        out.append(line("TOTAL", settings.currency + String(format: "%.2f", sale.total)))
        if vat { out.append(centre("VAT included / รวม VAT แล้ว")) }
        out.append(centre("Thank you / ขอบคุณค่ะ"))
        return out.joined(separator: "\n") + "\n"
    }
}

File: server/services/receipt.py
"""The receipt, in the two forms Thai law knows.

docs/research/01-legal-thailand.md section 4 and 04 section 5 (ledger H08, L09):

  VAT-registered shop (turnover > THB 1.8 m/yr)  ->  abbreviated tax invoice,
      Revenue Code s.86/6: the words "Tax Invoice (ABB)", seller name, address
      and taxpayer id, a running number, date, items, amount, and a line saying
      VAT is included.
  everyone else                                  ->  a plain receipt, no VAT lines.

Plain text, 32 columns, which is what an ESC/POS thermal printer takes as-is.
"""

from __future__ import annotations

from datetime import datetime

WIDTH = 32


def _line(left: str, right: str = "") -> str:
    left = left[:WIDTH - len(right) - 1] if right else left[:WIDTH]
    return f"{left:<{WIDTH - len(right)}}{right}" if right else left


def _centre(text: str) -> str:
    return text[:WIDTH].center(WIDTH).rstrip()


def render(sale: dict, settings: dict) -> str:
    """One sale row (as `Database.get_sale` returns it) to printable text."""
    vat = bool(settings.get("vat_registered"))
    cur = settings.get("currency", "฿")
    rate = float(settings.get("tax_rate", 0.07))
    when = datetime.fromisoformat(sale["timestamp"]).strftime("%d/%m/%Y %H:%M")

    out = [_centre(settings.get("store_name", "Store"))]
    if settings.get("store_address"):
        out.append(_centre(settings["store_address"]))
    if vat:
        out += [_centre("ใบกำกับภาษีอย่างย่อ"), _centre("TAX INVOICE (ABB)"),
                _centre(f"TIN {settings.get('tin', '') or '-'}")]
    else:
        out.append(_centre("ใบเสร็จรับเงิน / RECEIPT"))
    out += [_line(f"No. {sale['id']}"), _line(when), "-" * WIDTH]

    for it in sale["items"]:
        out.append(_line(it["product_name"]))
        out.append(_line(f"  {it['quantity']} x {it['price']:.2f}", f"{it['total']:.2f}"))

    out.append("-" * WIDTH)
    if vat:
        out.append(_line("Subtotal", f"{sale['subtotal']:.2f}"))
        out.append(_line(f"VAT {rate * 100:.0f}%", f"{sale['tax']:.2f}"))
    out.append(_line("TOTAL", f"{cur}{sale['total']:.2f}"))
    if vat:
        out.append(_centre("VAT included / รวม VAT แล้ว"))
    out.append(_centre("Thank you / ขอบคุณค่ะ"))
    return "\n".join(out) + "\n"
